### ISSUE TITLE
Modifier-only keybinds exist

### ISSUE THREAD

**Issue by miku4k** [niri-wm/niri#605]
You can set a modifier-only shortcut by using the xkb name for the modifier as the second key (e.g. Super+Super_L). Hitting just the modifier, or any keybind that includes the modifier, triggers the mod-only bind. 
I've only tested with Super+Super_L and the spawn action, but I believe other combinations behave similarly if not the same.

Checking that the non-mod key of a keybind isnt secretly a mod should be enough to fix this.

### System Information

* niri version: niri 0.1.7 (unknown commit)
(most recent version currently available in NixOS unstable)

* Distro: NixOS (unstable)

* GPU: AMD integrated

* CPU: AMD Ryzen 5 5500u

**Comment by YaLTeR** [comment]
Hm, I guess until we have release binds, this is more or less the best we can do. Though, maybe someone does use it, by carefully avoiding binds using that modifier...

**Comment by dev-nicolaos** [comment]
Is there an existing issue for release binds? I searched but couldn't find one. It would be nice to be able to bind an action to just Super without breaking all the other binds that use it as a modifier.

**Comment by YaLTeR** [comment]
I'm not sure there's an issue, but implementing release binds would involve some non trivial logic I imagine.

**Comment by jakesarjeant** [comment]
Is allowing mod-only binds necessarily a bug? I can think of at least a few use cases _for_ it, but I'm not clear on what problem it causes... Anyone who doesn't want mod-only binds can just not make any, but it doesn't seem like there's any reason to disallow them for people who want them (of course, release binds are kind of necessary for them to be useful). I'm also bouncing around the idea of showing a workspace overview (see #624) while the mod key is held instead of putting it in my bar, so it seems useful to bind opening the view to `Super+Super_L` with a timer to close it after a few seconds.

**Comment by miku4k** [comment]
they can lead to unexpected behavior if they arent accounted for, which they arent
especially the fact that they get executed

the disallowing them was more a temporary suggestion since i too think release binds would be complex to implement

and yes mod-only keybinds definitively have their uses (super opens some kind of menu on almost all desktops)

**Comment by jakesarjeant** [comment]
Maybe a better solution would be to introduce an `allow-mod-only;` option in the `binds` section of the config and disallow them otherwise?

**Comment by dev-nicolaos** [comment]
I'm having trouble imagining a mod only keybind being useful without also being a release bind. Wouldn't the bound action run on every invocation of every other shortcut that uses that mod key? If that's the case, I worry introducing an `allow-mod-only` option would just be a waste of time.

**Comment by miku4k** [comment]
yes they run every time you use that key be it in another shortcut or not.

**Comment by CodedNil** [comment]
Sorry if this is an obvious question, but does this mean that it's currently impossible to set for example super key alone to launch fuzzel? Or is there a hacky workaround to make that work until release binds are in?
Thanks.

**Comment by YaLTeR** [comment]
I don't think there's a workaround

**Comment by bbb651** [comment]
I have `lv3:caps_switch` set in `input.keyboard.xkb.options` to remap, and weirdly `ISO_Level3_Shift+ISO_Level3_Shift` is a valid keybind (I never realized this wasn't allowed with other modifier keys, so I guess it's a bug?)
So if you can find a way to remap <kbd>Super</kbd> to <kbd>ISO_Level3_Shift</kbd> with xkb options I think it should work.
(also I'm still on `0.1.10`, haven't tested it on `25.01`)

**Comment by YaLTeR** [comment]
> ISO_Level3_Shift+ISO_Level3_Shift is a valid keybind (I never realized this wasn't allowed with other modifier keys, so I guess it's a bug?

Uhh, I don't think this is intended. But I guess you can do it with other modifiers by using xkb names like `Super+Super_L`?

**Comment by boomskats** [comment]
@CodedNil I use key overloading in [keyd](https://github.com/rvaiya/keyd) for mod-tap like behaviour when remapping caps to ctrl/esc. I've found it to be reliable - should be pretty trivial to use it to do what you're wanting to do.

**Comment by CodedNil** [comment]
> [@CodedNil](https://github.com/CodedNil) I use key overloading in [keyd](https://github.com/rvaiya/keyd) for mod-tap like behaviour when remapping caps to ctrl/esc. I've found it to be reliable - should be pretty trivial to use it to do what you're wanting to do.

That's amazing thanks dude!!!!! This works pretty flawlessly for me until release binds are in.

My nix config if it helps anyone sort similar
services.keyd = {
  enable = true;
  keyboards.default = {
    ids = [ "*" ];
    settings.main = {
      compose = "layer(meta)"; # Make the menu key press super
      leftmeta = "overload(meta, macro(leftmeta+z))"; # Make left meta tap open anyrun keybind
    };
  };
};

**Comment by tmarkov** [comment]
> I'm having trouble imagining a mod only keybind being useful without also being a release bind. Wouldn't the bound action run on every invocation of every other shortcut that uses that mod key? If that's the case, I worry introducing an `allow-mod-only` option would just be a waste of time.

I have this use case:
Press Control -> Enable microphone
Release Control -> Disable microphone

Which essentially implements push to talk for me. Of course, this would short-term enable mic when copy/pasting, for example, but that's acceptable side-effect.

**Comment by miku4k** [comment]
im curious, how do you disable the microphone on release? 

also relevant xkcd <https://xkcd.com/1172/>

**Comment by tmarkov** [comment]
Not with Niri. The inability to have some version of push-to-talk is why I'm not using Niri currently. I'm simply demonstrating a use-case for mod-only-keybind that's not a release bind.

As for the xkcd, I fail to see the relevance. AFAIK this doesn't rely on a bug, and is usually the simplest way to achieve push-to-talk (only hyprland has something better).

**Comment by YaLTeR** [comment]
I see. I guess for this specific microphone push-to-talk case, some system, for forwarding keys to specific windows would also work.

**Comment by shadowe1ite** [comment]
Any updates?!

**Comment by tacoearth** [comment]
Would be nice if we get a modifier for long press, in my case I use SUPER for fuzzel

**Comment by sagebind** [comment]
Would like to see this feature as well. For now I did a similar workaround as the one recommended using keyd, except using xremap instead for whoever this may help:

```yaml
modmap:
  - name: Overview when Super is pressed alone
    remap:
      LeftMeta:
        held: LeftMeta
        alone:
          - LeftMeta
          - O
        free_hold: true
```

With this, just pressing and releasing Meta/Mod by itself will send Mod+O to Niri instead, so you can have that bound to whatever (in my case, `toggle-overview`).

**Comment by Hate9** [comment]
In case it helps someone else, the *best* solution I found, to both allow this *and* not mess with other DEs (based on some suggestions here and on my own searching), was:
- set keyd's default config to just
  ```
  [ids]
  *
  ```
  (so it applies to all keyboards)
- enable keyd's systemd unit (obvious to some people, but just in case it isn't obvious to whoever's reading this)
- add the user to whatever group owns `/var/run/keyd.socket` (on my system it was `keyd`)
- add `spawn-at-startup "keyd bind 'leftmeta=timeout(overload(meta,macro(leftmeta+z)),100,leftmeta)'"` to your niri config (so it only applies inside niri) (assuming your modkey is `meta` - change this if not)
  - (the `timeout` bit means it won't activate after you, for instance, resize a window with mod+rightclick+drag)
- use `Mod+Z` for the bind you want to set (e.g. `Mod+Z { spawn "fuzzel"; }`)

**Comment by gevera** [comment]
> In case it helps someone else, the _best_ solution I found, to both allow this _and_ not mess with other DEs (based on some suggestions here and on my own searching), was:
> 
>     * set keyd's default config to just
>       ```
>       [ids]
>       *
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       (so it applies to all keyboards)
> 
>     * enable keyd's systemd unit (obvious to some people, but just in case it isn't obvious to whoever's reading this)
> 
>     * add the user to whatever group owns `/var/run/keyd.socket` (on my system it was `keyd`)
> 
>     * add `spawn-at-startup "keyd bind 'leftmeta=timeout(overload(meta,macro(leftmeta+z)),100,leftmeta)'"` to your niri config (so it only applies inside niri) (assuming your modkey is `meta` - change this if not)
>       
>       * (the `timeout` bit means it won't activate after you, for instance, resize a window with mod+rightclick+drag)
> 
>     * use `Mod+Z` for the bind you want to set (e.g. `Mod+Z { spawn "fuzzel"; }`)

Can you elaborate on this? Seems like what I am looking for

**Comment by pyluyten** [comment]
I also tried @Hate9 solution but even using an empty keyd config and using "keyd bind" under Niri , the bind is still active after logging to another compositor. I could have missed something
